bentoml/_internal/server/grpc/server.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+
+import typing as t
+import asyncio
+import logging
+from typing import TYPE_CHECKING
+
+import grpc
+from grpc import aio
+
+from ...utils import LazyLoader
+from ...utils import cached_property
+
+logger = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from grpc_health.v1 import health_pb2 as pb_health
+    from grpc_health.v1 import health_pb2_grpc as services_health
+    from typing_extensions import Self
+
+    from bentoml.grpc.v1alpha1 import service_pb2_grpc as services
+
+    from .config import Config
+else:
+    from bentoml.grpc.utils import import_generated_stubs
+
+    _, services = import_generated_stubs()
+    health_exception_msg = "'grpcio-health-checking' is required for using health checking endpoints. Install with 'pip install grpcio-health-checking'."
+    pb_health = LazyLoader(
+        "pb_health",
+        globals(),
+        "grpc_health.v1.health_pb2",
+        exc_msg=health_exception_msg,
+    )
+    services_health = LazyLoader(
+        "services_health",
+        globals(),
+        "grpc_health.v1.health_pb2_grpc",
+        exc_msg=health_exception_msg,
+    )
+
+
+class Server:
+    """An async implementation of a gRPC server."""
+
+    def __init__(self, config: Config):
+        self.config = config
+        self.servicer = config.servicer
+        self.loaded = False
+
+    @cached_property
+    def loop(self) -> asyncio.AbstractEventLoop:
+        return asyncio.get_event_loop()
+
+    def load(self) -> Self:
+        from concurrent.futures import ThreadPoolExecutor
+
+        assert not self.loaded
+        if not bool(self.servicer):
+            self.servicer.load()
+        assert self.servicer.loaded
+
+        self.server = aio.server(
+            migration_thread_pool=ThreadPoolExecutor(
+                max_workers=self.config.migration_thread_pool_workers
+            ),
+            options=self.config.options,
+            maximum_concurrent_rpcs=self.config.maximum_concurrent_rpcs,
+            handlers=self.config.handlers,
+            interceptors=self.servicer.interceptors_stack,
+        )
+        self.loaded = True
+
+        return self
+
+    def run(self) -> None:
+        if not self.loaded:
+            self.load()
+        assert self.loaded
+
+        try:
+            self.loop.run_until_complete(self.serve())
+        finally:
+            try:
+                self.loop.call_soon_threadsafe(
+                    lambda: asyncio.ensure_future(self.shutdown())
+                )
+            except Exception as e:  # pylint: disable=broad-except
+                raise RuntimeError(
+                    f"Server failed unexpectedly. enable GRPC_VERBOSITY=debug for more information: {e}"
```

Comment:
Should we do these ourselves if `BENTOML_DEBUG=true`?

Reply:
I think we shouldn't enable GRPC_VERBOSITY with `--debug` cuz it is a lot of details. I think if user ever run into this issue, they can do it themselves.

Reply:
Isn't that the point of `--debug`, though?

Reply:
ok let me do that.